### PR TITLE
Add playtime command

### LINKED ISSUE #3827: Add a way to see the amount of time another player has played on the server, e.g. /playtime (player)
### Feature request

**Feature description:**  

The feature I'm suggesting is a way to see how much time another player has played in total on the server, possibly with a config option that disables counting AFK time.

**How the feature is useful:**  

This feature is useful because it allows server owners to check how much a player has played without using plugins aside from Essentials. It would be very convenient to have this.

### PR DESCRIPTION
**Pull request by melluh**
### Information

This PR closes #3827. 

### Details

**Proposed feature:**  

Allow players to view their own or another player's playtime.

**Environments tested:**    

OS: Windows 10

Java version: ``java version "16.0.1" 2021-04-20``

- [x] Most recent Paper version (1.17.1, git-Paper-fdd5c65)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:**

### CODE REVIEW

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.bukkit.Statistic;
+
+import com.earth2me.essentials.CommandSource;
+import com.earth2me.essentials.User;
+import com.earth2me.essentials.utils.DateUtil;
+import com.earth2me.essentials.utils.EnumUtil;
+
+public class Commandplaytime extends EssentialsCommand {
+
+    // For some reason, in 1.13 PLAY_ONE_MINUTE = ticks played = what used to be PLAY_ONE_TICK
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
    // For some reason, in 1.13 PLAY_ONE_MINUTE = ticks played = what used to be PLAY_ONE_TICK
```

File: Essentials/src/main/resources/messages_en.properties [hunk 2]
```diff
@@ -1481,3 +1481,6 @@ years=years
 youAreHealed=§6You have been healed.
 youHaveNewMail=§6You have§c {0} §6messages\! Type §c/mail read§6 to view your mail.
 xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
+playtime=§6{0}§6''s Playtime\:§r {1}
```
Comment (JRoy) [inline Essentials/src/main/resources/messages_en.properties]: Do not add messages to this file, add them to `messages.properties`, not `messages_en.properties`.
```suggestion
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            throw new NotEnoughArgumentsException();
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
+    }
+    
+    @Override
+    protected List<String> getTabCompleteOptions(final Server server, final CommandSource sender, final String commandLabel, final String[] args) {
+        if (args.length == 1) {
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
        if (args.length == 1 && sender.isAuthorized("essentials.playtime.others", ess)) {
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        if (args.length >= 1) {
+            target = this.getPlayer(server, sender, args[0]);
+            if (target == null) {
+                throw new PlayerNotFoundException();
+            }
+        } else if (sender.isPlayer()) {
+            target = ess.getUser(sender.getPlayer());
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
        if (args.length > 0) {
            target = getPlayer(server, sender, args, 0, false, false);
        } else if (sender.isPlayer()) {
            target = sender.getUser(ess);
        } else {
            throw new NotEnoughArgumentsException();
        }
```

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/resources/messages.properties [hunk 5]
```diff
@@ -1491,3 +1491,6 @@ years=years
 youAreHealed=\u00a76You have been healed.
 youHaveNewMail=\u00a76You have\u00a7c {0} \u00a76messages\! Type \u00a7c/mail read\u00a76 to view your mail.
 xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
+playtime=\u00a76{0}\u00a76''s Playtime\:\u00a7r {1}
```
Comment (JRoy) [inline Essentials/src/main/resources/messages.properties]: please add these in alphabetical order

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/resources/messages_en.properties [hunk 6]
```diff
@@ -1480,4 +1480,4 @@ year=year
 years=years
 youAreHealed=§6You have been healed.
 youHaveNewMail=§6You have§c {0} §6messages\! Type §c/mail read§6 to view your mail.
-xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
+xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.
```
Comment (JRoy) [inline Essentials/src/main/resources/messages_en.properties]:
```suggestion
xmppNotConfigured=XMPP is not configured properly. If you do not know what XMPP is, you may wish to remove the EssentialsXXMPP plugin from your server.

```

File: Essentials/src/main/resources/messages.properties [hunk 7]
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
 playerUnbanIpAddress=\u00a76Player\u00a7c {0} \u00a76unbanned IP\:\u00a7c {1}
 playerUnbanned=\u00a76Player\u00a7c {0} \u00a76unbanned\u00a7c {1}
 playerUnmuted=\u00a76You have been unmuted.
+playtimeCommandDescription=Shows a player''s playtime
+playtimeCommandUsage=/<command> [player]
```
Comment (JRoy) [inline Essentials/src/main/resources/messages.properties:888]:
```suggestion
playtimeCommandUsage=/<command> [player]
playtimeCommandUsage1=/<command>
playtimeCommandUsage1Description=Shows your time played in game
playtimeCommandUsage2=/<command> <player>
playtimeCommandUsage2Description=Shows the specified player''s time played in game
```

File: Essentials/src/main/resources/messages.properties [hunk 8]
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
 playerUnbanIpAddress=\u00a76Player\u00a7c {0} \u00a76unbanned IP\:\u00a7c {1}
 playerUnbanned=\u00a76Player\u00a7c {0} \u00a76unbanned\u00a7c {1}
 playerUnmuted=\u00a76You have been unmuted.
+playtimeCommandDescription=Shows a player''s playtime
```
Comment (JRoy) [inline Essentials/src/main/resources/messages.properties]:
```suggestion
playtimeCommandDescription=Shows a player''s time played in game
```

File: Essentials/src/main/resources/plugin.yml [hunk 9]
```diff
@@ -344,6 +344,10 @@ commands:
     description: Pong!
     usage: /<command>
     aliases: [echo,eecho,eping,pong,epong]
+  playtime:
+    description: Shows a player's playtime
```
Comment (JRoy) [inline Essentials/src/main/resources/plugin.yml]:
```suggestion
    description: Shows a player's time played in game
```

File: Essentials/src/main/resources/messages.properties [hunk 10]
```diff
@@ -872,6 +872,9 @@ playerTempBanned=\u00a76Player \u00a7c{0}\u00a76 temporarily banned \u00a7c{1}\u
...
+playtimeCommandUsage=/<command> [player]
+playtime=\u00a76{0}\u00a76''s Playtime\:\u00a7r {1}
```
Comment (JRoy) [inline Essentials/src/main/resources/messages.properties]:
```suggestion
playtime=\u00a76Playtime\:\u00a7c {0}
playtimeOther=\u00a76Playtime of {1}\u00a76\:\u00a7c {0}
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        if (args.length > 0 && sender.isAuthorized("essentials.playtime.others", ess)) {
+            target = getPlayer(server, sender, args, 0);
+        } else if (sender.isPlayer()) {
+            target = sender.getUser(ess);
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
        final IUser target;
        final String key;
        
        if (args.length > 0 && sender.isAuthorized("essentials.playtime.others", ess)) {
            target = getPlayer(server, sender, args, 0);
            key = "playtimeOther";
        } else if (sender.isPlayer()) {
            target = sender.getUser(ess);
            key = "playtime";
        } else {
            throw new NotEnoughArgumentsException();
        }

        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```

**Review by JRoy — CHANGES_REQUESTED** [review]

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+            key = "playtimeOther";
+        } else if (sender.isPlayer()) {
+            target = sender.getUser(ess);
+            key = "playtime";
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl("playtime", target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
        sender.sendMessage(tl(key, target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+            key = "playtimeOther";
+        } else if (sender.isPlayer()) {
+            target = sender.getUser(ess);
+            key = "playtime";
+        } else {
+            throw new NotEnoughArgumentsException();
+        }
+
+        final long playtimeMs = System.currentTimeMillis() - (target.getBase().getStatistic(PLAY_ONE_TICK) * 50);
+        sender.sendMessage(tl(key, target.getDisplayName(), DateUtil.formatDateDiff(playtimeMs)));
```
Comment (JRoy) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandplaytime.java]:
```suggestion
        sender.sendMessage(tl(key, DateUtil.formatDateDiff(playtimeMs), target.getDisplayName()));
```

**Review by JRoy — APPROVED** [review]

**Comment by ghost** [comment]
@MisterMelDev, would it be possible to add offline playtime support? This would be insanely useful.

**Comment by melluh** [comment]
> @MisterMelDev, would it be possible to add offline playtime support? This would be insanely useful.

The way I implemented this, it currently wouldn't be possible. I'm using the built-in Bukkit statistics system, which also contains a statistic for playtime, but this can only be retrieved when the player is online. The plugin would have to store the playtime in the player data or something. I'm a bit busy currently, but maybe someone else could make a PR for that?

**Comment by serious-angel** [comment]
> The way I implemented this

Greetings! Just in case if still interested in offline data, there are a few implementation like the following wonderful one for a reference:
- <https://www.spigotmc.org/resources/playtime.26016>;
- <https://github.com/Twi5TeD/PlayTime>.

**Pull request merged**